File: DecorLayer.cpp
#include "DecorLayer.h"
#include <random>     // For std::mt19937, distributions
#include <cmath>      // For std::cos, std::sin
#include <ctime>      // For std::time used in Rng seed
#include "Utils.h"




//--------------------------------------------------------------------
//  DecorLayer Implementation
//--------------------------------------------------------------------

// Constructor
DecorLayer::DecorLayer(std::size_t target) : m_targetCount(target) {}

// Update method
void DecorLayer::update(float dt, sf::Vector2u winSize, const ColorTheme& theme) {
    while (m_shapes.size() < m_targetCount) {
        m_spawn(winSize, theme.decorBase, theme.decorAccent1, theme.decorAccent2);
    }

    for (auto& s : m_shapes) {
        s.pos += s.vel * dt;
        // Simplified wrapping logic
        if (s.vel.x > 0 && s.pos.x - s.size > static_cast<float>(winSize.x)) {
            s.pos.x = -s.size - randRange<float>(50.f, 150.f); // Start further left
            s.pos.y = randRange<float>(0.f, static_cast<float>(winSize.y));
        }
        else if (s.vel.x < 0 && s.pos.x + s.size < 0.f) {
            s.pos.x = static_cast<float>(winSize.x) + s.size + randRange<float>(50.f, 150.f); // Start further right
            s.pos.y = randRange<float>(0.f, static_cast<float>(winSize.y));
        }
        // Simple vertical wrap (adjust if needed)
        if (s.pos.y - s.size > static_cast<float>(winSize.y)) s.pos.y = -s.size;
        else if (s.pos.y + s.size < 0.f) s.pos.y = static_cast<float>(winSize.y) + s.size;
    }
}

// Draw method
void DecorLayer::draw(sf::RenderTarget& rt) const {
    for (const auto& s : m_shapes) {
        switch (s.kind) {
        case Kind::Circle: {
            sf::CircleShape c(s.size);
            c.setOrigin({ s.size, s.size });
            c.setPosition(s.pos);
            c.setFillColor(s.col);
            rt.draw(c);
            break;
        }
        case Kind::Triangle: {
            sf::ConvexShape tri(3);
            sf::Vector2f centerSum(0, 0);
            for (int i = 0; i < 3; ++i) {
                float ang = DEG2RAD(s.rot + i * 120.f); // Use renamed helper
                sf::Vector2f point = { s.size * std::cos(ang), s.size * std::sin(ang) };
                tri.setPoint(i, point);
                centerSum += point;
            }
            tri.setOrigin(centerSum / 3.f);
            tri.setPosition(s.pos);
            tri.setFillColor(s.col);
            rt.draw(tri);
            break;
        }
        case Kind::Line: {
            sf::RectangleShape line({ s.size, 2.f }); // Slightly thicker line
            line.setOrigin({ s.size / 2.f, 1.f }); // Center origin
            line.setPosition(s.pos);
            line.setRotation(sf::degrees(s.rot)); // Use sf::degrees
            line.setFillColor(s.col);
            rt.draw(line);
            break;
        }
        }
    }
}

// Spawn method (made private)
void DecorLayer::m_spawn(sf::Vector2u win, const sf::Color& base, const sf::Color& accent1, const sf::Color& accent2) {
    Shape s;
    bool spawnLeft = randRange(0, 1) == 0;

    s.kind = static_cast<Kind>(randRange<int>(0, 2));
    float yPos = randRange<float>(0.f, static_cast<float>(win.y));
    float xPos = spawnLeft ? -randRange<float>(50.f, 200.f) : static_cast<float>(win.x) + randRange<float>(50.f, 200.f);
    s.pos = { xPos, yPos };
    s.vel = { (spawnLeft ? 1.f : -1.f) * randRange<float>(15.f, 45.f), 0.f };
    s.rot = randRange<float>(0.f, 360.f);

    switch (s.kind) {
    case Kind::Circle: // Use Base color
        s.size = randRange<float>(50.f, 120.f);
        s.col = sf::Color(base.r, base.g, base.b, randRange<int>(60, 100));
        break;
    case Kind::Triangle: // Use Accent1 color
        s.size = randRange<float>(40.f, 150.f);
        s.col = sf::Color(accent1.r, accent1.g, accent1.b, randRange<int>(30, 60));
        break;
    case Kind::Line: // Use Accent2 color
        s.size = randRange<float>(80.f, 180.f);
        s.col = sf::Color(accent2.r, accent2.g, accent2.b, randRange<int>(80, 140));
        break;
    }
    m_shapes.push_back(s); // Add to member vector
}

File: web/src/render/decorLayer.ts
import type { Vec2, Color } from "../core/types";
import type { ColorTheme } from "../core/themes";
import { colorToCss, randRange } from "../core/utils";
import { degToRad } from "../core/utils";

export enum DecorKind {
  Circle = "Circle",
  Triangle = "Triangle",
  Line = "Line"
}

interface DecorShape {
  kind: DecorKind;
  pos: Vec2;
  vel: Vec2;
  size: number;
  rot: number;
  color: Color;
}

export class DecorLayer {
  private shapes: DecorShape[] = [];

  constructor(private readonly targetCount = 25) {}

  update(dt: number, winSize: Vec2, theme: ColorTheme) {
    while (this.shapes.length < this.targetCount) {
      this.spawn(winSize, theme.decorBase, theme.decorAccent1, theme.decorAccent2);
    }

    for (const shape of this.shapes) {
      shape.pos = { x: shape.pos.x + shape.vel.x * dt, y: shape.pos.y + shape.vel.y * dt };
      if (shape.vel.x > 0 && shape.pos.x - shape.size > winSize.x) {
        shape.pos.x = -shape.size - randRange(50, 150);
        shape.pos.y = randRange(0, winSize.y);
      } else if (shape.vel.x < 0 && shape.pos.x + shape.size < 0) {
        shape.pos.x = winSize.x + shape.size + randRange(50, 150);
        shape.pos.y = randRange(0, winSize.y);
      }
      if (shape.pos.y - shape.size > winSize.y) shape.pos.y = -shape.size;
      else if (shape.pos.y + shape.size < 0) shape.pos.y = winSize.y + shape.size;
    }
  }

  draw(ctx: CanvasRenderingContext2D) {
    for (const shape of this.shapes) {
      ctx.save();
      ctx.fillStyle = colorToCss(shape.color);

      if (shape.kind === DecorKind.Circle) {
        ctx.beginPath();
        ctx.arc(shape.pos.x, shape.pos.y, shape.size, 0, Math.PI * 2);
        ctx.fill();
      } else if (shape.kind === DecorKind.Triangle) {
        const points: Vec2[] = [];
        for (let i = 0; i < 3; i += 1) {
          const ang = degToRad(shape.rot + i * 120);
          points.push({
            x: shape.pos.x + Math.cos(ang) * shape.size,
            y: shape.pos.y + Math.sin(ang) * shape.size
          });
        }
        ctx.beginPath();
        ctx.moveTo(points[0].x, points[0].y);
        ctx.lineTo(points[1].x, points[1].y);
        ctx.lineTo(points[2].x, points[2].y);
        ctx.closePath();
        ctx.fill();
      } else if (shape.kind === DecorKind.Line) {
        ctx.translate(shape.pos.x, shape.pos.y);
        ctx.rotate(degToRad(shape.rot));
        ctx.fillRect(-shape.size / 2, -1, shape.size, 2);
      }

      ctx.restore();
    }
  }

  private spawn(win: Vec2, base: Color, accent1: Color, accent2: Color) {
    const spawnLeft = randRange(0, 1, true) === 0;
    const kindIndex = randRange(0, 2, true);
    const kind = [DecorKind.Circle, DecorKind.Triangle, DecorKind.Line][kindIndex] ?? DecorKind.Circle;
    const yPos = randRange(0, win.y);
    const xPos = spawnLeft ? -randRange(50, 200) : win.x + randRange(50, 200);

    const shape: DecorShape = {
      kind,
      pos: { x: xPos, y: yPos },
      vel: { x: (spawnLeft ? 1 : -1) * randRange(15, 45), y: 0 },
      size: 0,
      rot: randRange(0, 360),
      color: base
    };

    if (kind === DecorKind.Circle) {
      shape.size = randRange(50, 120);
      shape.color = { ...base, a: randRange(60, 100, true) };
    } else if (kind === DecorKind.Triangle) {
      shape.size = randRange(40, 150);
      shape.color = { ...accent1, a: randRange(30, 60, true) };
    } else {
      shape.size = randRange(80, 180);
      shape.color = { ...accent2, a: randRange(80, 140, true) };
    }

    this.shapes.push(shape);
  }
}
